### ISSUE TITLE
Feature request: Allow for custom headers

### ISSUE THREAD

**Issue by ldesgoui** [jahewson/elm-graphql#30]
Hello,

This tool would be great but it only exposes Tasks and is therefore impossible to use in the case of a header-based authentication. Do you think it could be possible to decouple or allow for custom headers ?

Thanks

**Comment by jahewson** [comment]
Yes, that would be possible. Do you have a preference?

**Comment by ldesgoui** [comment]
Not really, but I think that GraphQL is meant to be independent of the HTTP implementation, I feel the tools should be as well.
Best case scenario would be letting people implement their own fetcher, similar to what graphiql does (https://github.com/graphql/graphiql#usage), with a function of same/similar type as GraphQL.query/GraphQL.mutation, something along the lines of:

    queryFriendsAuthed : Token -> { id : String } -> Task Http.Error QueryFriendsResult
    queryFriendsAuthed token =
        GraphQL.Starwars.queryFriendsGeneric (bearerQuery token)

    bearerQuery : Token -> Url -> Query -> Params -> Decoder a -> Task Http.Error a

    init { token } =
        initialModel
            ! [ queryFriendsAuthed token { id = 0 }
                    |> Task.perform Ok Err
              ]


Thanks

**Comment by ldesgoui** [comment]
q1Generic : (String -> String -> String -> String -> Json.Encode.Value -> Decoder a -> Task e a) -> { a : Maybe String } -> Task e Q1
    q1Generic fetcher params =
        let
            graphQLQuery =
                """query q1($a: String) { test1 { __typename ... on List { name } ... on Task { text } } }"""

            graphQLParams =
                Json.Encode.object
                    [ ( "a", Json.Encode.string params.a )
                    ]
        in
            fetcher "query" endpointUrl graphQLQuery "q1" graphQLParams q1Decoder


    q1 : { a : Maybe String } -> Task Http.Error Q1
    q1 params =
        q1Generic (\_ -> GraphQL.query) params

How does that look ? I've yet to compile using it, but I'm pretty sure it should be alright.

**Comment by aaronwhite** [comment]
Customer headers and/or HTTP independence (via a functional argument, etc) would be super useful.... this library is great, and I'd love to incorporate it, but I need to hack the output to insert my auth headers at the moment
